cloud_functions/climateiq_data_export_cf/main.py

```diff
@@ -211,19 +228,173 @@ def _build_spatialized_model_predictions(
+    chunk_metadata: dict,
+    spatialized_predictions: pd.DataFrame,
+) -> pd.Series:
+    """Projects cell centroids to H3 indexes.
+
+    Filters out cells that have projected H3 centroids outside of the current
+    chunk and de-dupes duplicate H3 projections by using the average prediction
+    value from overlapping cells.
+
+    Args:
+        study_area_metadata: A dictionary containing metadata for the study
+        area.
+        chunk_metadata: A dictionary containing metadata for the chunk.
+        spatialized_predictions: A DataFrame containing the lat/lon coordinates of cell
+        centroids in a single chunk along with associated predictions
+
+    Returns:
+        A Series containing H3 indexes in a single chunk along with associated
+        predictions, representing a subset of the full study area results.
+    """
+    # Calculate H3 information for each cell.
+    spatialized_predictions[
+        ["h3_index", "h3_centroid_lat", "h3_centroid_lon", "h3_boundary"]
+    ] = spatialized_predictions.apply(_add_h3_index_details, axis=1)
+
+    # Filter out any rows where the projected H3 centroid falls outside of the
+    # chunk boundary.
+    chunk_boundary = _get_chunk_boundary(study_area_metadata, chunk_metadata)
+    spatialized_predictions = spatialized_predictions[
+        spatialized_predictions.apply(
+            lambda cell: chunk_boundary.contains(
+                geometry.Point(cell["h3_centroid_lon"], cell["h3_centroid_lat"])
+            ),
+            axis=1,
+        )
+    ]
+
+    # Extract rows where the projected H3 cell is not fully contained within the chunk
+    # so we can aggregate prediction values across chunk boundaries.
+    boundary_h3_cells = (
+        spatialized_predictions[
+            spatialized_predictions.apply(
+                lambda row: not row["h3_boundary"].within(chunk_boundary),
+                axis=1,
+            )
+        ]
+        .groupby(["h3_index"])
+        .prediction.agg("mean")
+    )
+
+    # TODO: Add logic for fetching predictions from neighboring chunks for boundary
+    # cells.
+    spatialized_predictions = spatialized_predictions.groupby(
+        ["h3_index"]
+    ).prediction.agg("mean")
+    return spatialized_predictions
+
+
+def _get_neighboring_chunks(
+    study_area_metadata: dict, chunk_metadata: dict, boundary_h3_cells: pd.DataFrame
+):
+    x = chunk_metadata["x_index"]
+    y = chunk_metadata["y_index"]
+
+    neighbors = {(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)}
+    queue = queue.Queue()
+    for neighbor in neighbors:
+        queue.put(neighbor)
+
+    while not queue.empty():
+        neighbor_x, neighbor_y = queue.get()
+        query = (
+            chunks_ref.where("x_index", "==", neighbor_x)
+            .where("y_index", "==", neighbor_y)
+            .limit(1)
+        )
+        chunk_doc = query.get()
+        if not chunk_doc.exists:
+            # Some chunks may not have neighbors in the study area.
```

Comment:
would we handle those two use cases separately? if there is an error for a chunk, should we stop the entire pipeline? if not, I would prefer the simpler logic of just checking whether or not the chunk exists regardless of the reason

Reply:
Yes, we'd need to handle that separately. We need to know if we're missing data or not before we export!
Fine to do this in future PRs, but we should note that we need to do this.

Reply:
We also need to think about how we're triggering this cloud function. Do we know for sure if the neighbors will already be written to GCS or not? If a neighbor is not present because it hasn't been written yet, as opposed to a neighbor that's out of bounds and will never be written, that's something we need to handle!

Reply:
Can you add more context on what we'd do if there is missing data?

Also regarding triggering of this function- the USL would like to manually trigger this pipeline once verification of CNN outputs is complete. Still need to look into manual triggers but this alleviates the problems of some neighbors not being ready yet if the function was triggered automatically

Reply:
> Can you add more context on what we'd do if there is missing data?

Raise an error & not export data. We don't want to export data with holes in it!

Reply:
Makes sense, updated to through an error